src/audio/mixin_mixout/mixin_mixout.c

```diff
@@ -481,88 +208,67 @@ static int mixout_process(struct processing_module *mod,
+	uint32_t j = 0;
+	int i;
 
-		source_index = find_module_source_index(mod_source_info, source);
-		/* this shouldn't happen but skip even if it does and move to the next source */
-		if (source_index < 0)
+	/* mix streams */
+	for (i = 0; i < num_input_buffers; i++) {
+		if (input_buffers[i].size == 0)
 			continue;
 
-		pending_frames = md->pending_frames[source_index];
-
-		if (source->state == COMP_STATE_ACTIVE || pending_frames)
-			frames_to_produce = MIN(frames_to_produce, pending_frames);
+		sources_stream[j] = mod->input_buffers[i].data;
+		sources_indices[j++] = i;
+		frames = MIN(frames, input_buffers[i].size);
 	}
 
-	if (frames_to_produce > 0 && frames_to_produce < INT32_MAX) {
-		for (i = 0; i < num_input_buffers; i++) {
-			const struct audio_stream __sparse_cache *source_stream;
-			struct comp_buffer __sparse_cache *unused_in_between_buf;
-			struct comp_dev *source;
-			int source_index;
-			uint32_t pending_frames;
-
-			source_stream = input_buffers[i].data;
-			unused_in_between_buf = attr_container_of(source_stream,
-								  struct comp_buffer __sparse_cache,
-								  stream, __sparse_cache);
-
-			source = unused_in_between_buf->source;
-
-			source_index = find_module_source_index(mod_source_info, source);
-			if (source_index < 0)
-				continue;
-
-			pending_frames = md->pending_frames[source_index];
-			if (pending_frames >= frames_to_produce)
-				md->pending_frames[source_index] -= frames_to_produce;
-			else
-				md->pending_frames[source_index] = 0;
-		}
+	/* generate silence if no data available from any of the sources */
+	if (!j) {
+		/*
+		 * Generate silence when sources are inactive. When
+		 * sources change to active, additionally keep
+		 * generating silence until at least one of the
+		 * sources start to have data available (frames!=0).
+		 */
+		sink_bytes =
+			mod->dev->frames * audio_stream_frame_bytes(mod->output_buffers[0].data);
+		if (!audio_stream_set_zero(mod->output_buffers[0].data, sink_bytes))
+			mod->output_buffers[0].size = sink_bytes;
 
-		assert(md->mixed_frames >= frames_to_produce);
-		md->mixed_frames -= frames_to_produce;
-
-		sink_bytes = frames_to_produce *
-				audio_stream_frame_bytes(output_buffers[0].data);
-		output_buffers[0].size = sink_bytes;
-	} else {
-		sink_bytes = dev->frames * audio_stream_frame_bytes(output_buffers[0].data);
-		if (!audio_stream_set_zero(output_buffers[0].data, sink_bytes))
-			output_buffers[0].size = sink_bytes;
-		else
-			output_buffers[0].size = 0;
+		return 0;
 	}
 
-	module_source_info_release(mod_source_info);
+	md->mix_func(mod->dev, mod->output_buffers[0].data, sources_stream, j, frames);
+
+	sink_bytes = frames * audio_stream_frame_bytes(mod->output_buffers[0].data);
+	mod->output_buffers[0].size = sink_bytes;
+
+	/* update source buffer consumed bytes */
+	for (i = 0; i < j; i++) {
+		source_bytes = frames * audio_stream_frame_bytes(mod->input_buffers[i].data);
+		mod->input_buffers[sources_indices[i]].consumed = source_bytes;
```

Comment:
since the mix function is the mixing the sources with same format & channels, so how about just use the sink_bytes and add comment to explain it？